// @mxenabled/react-native-widget-sdk/dist/src/sso/environment.js
export var Environment;
(function (Environment) {
    Environment["SAND"] = "sand";
    Environment["QA"] = "qa";
    Environment["INT"] = "integration";
    Environment["PROD"] = "production";
})(Environment || (Environment = {}));
const environmentToEnvironment = {
    [Environment.SAND]: Environment.SAND,
    [Environment.QA]: Environment.QA,
    [Environment.INT]: Environment.INT,
    [Environment.PROD]: Environment.PROD,
};
export function lookupEnvironment(str) {
    const env = environmentToEnvironment[str];
    if (!env) {
        throw new Error(`Invalid MX environment: ${str}`);
    }
    return env;
}
const environmentToHost = {
    [Environment.SAND]: "https://api.sand.internal.mx",
    [Environment.QA]: "https://api.qa.internal.mx",
    [Environment.INT]: "https://int-api.mx.com",
    [Environment.PROD]: "https://api.mx.com",
};
export function lookupHost(str) {
    const env = lookupEnvironment(str);
    const host = environmentToHost[env];
    if (!host) {
        throw new Error(`Missing host for MX environment: ${str}`);
    }
    return host;
}
