// @mxenabled/react-native-widget-sdk/dist/src/sso/widget_configuration.js
/**
 * See "Configuration options" section in
 * https://docs.mx.com/api#connect_request_a_url for more details.
 */
export var Type;
(function (Type) {
    Type["AccountsWidget"] = "accounts_widget";
    Type["BudgetsWidget"] = "budgets_widget";
    Type["ConnectWidget"] = "connect_widget";
    Type["ConnectionsWidget"] = "connections_widget";
    Type["DebtsWidget"] = "debts_widget";
    Type["FinstrongWidget"] = "finstrong_widget";
    Type["GoalsWidget"] = "goals_widget";
    Type["HelpWidget"] = "help_widget";
    Type["MasterWidget"] = "master_widget";
    Type["MiniBudgetsWidget"] = "mini_budgets_widget";
    Type["MiniFinstrongWidget"] = "mini_finstrong_widget";
    Type["MiniPulseCarouselWidget"] = "mini_pulse_carousel_widget";
    Type["MiniSpendingWidget"] = "mini_spending_widget";
    Type["PulseWidget"] = "pulse_widget";
    Type["SettingsWidget"] = "settings_widget";
    Type["SpendingWidget"] = "spending_widget";
    Type["TransactionsWidget"] = "transactions_widget";
    Type["TrendsWidget"] = "trends_widget";
})(Type || (Type = {}));
export function getWidgetConfigurationFromProps(props) {
    return {
        client_redirect_url: props.clientRedirectUrl,
        color_scheme: props.colorScheme,
        current_institution_code: props.currentInstitutionCode,
        current_institution_guid: props.currentInstitutionGuid,
        current_member_guid: props.currentMemberGuid,
        disable_institution_search: props.disableInstitutionSearch,
        include_transactions: props.includeTransactions,
        is_mobile_webview: true,
        mode: props.mode,
        ui_message_version: 4,
        ui_message_webview_url_scheme: props.uiMessageWebviewUrlScheme,
        update_credentials: props.updateCredentials,
        widget_type: props.widgetType,
    };
}
